File: src/test/java/chat/teco/tecochat/chat/application/ChatQueryServiceTest.java
package chat.teco.tecochat.chat.application;

import static chat.teco.tecochat.chat.domain.Answer.answer;
import static chat.teco.tecochat.chat.domain.Question.question;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;

import chat.teco.tecochat.chat.domain.Answer;
import chat.teco.tecochat.chat.domain.Chat;
import chat.teco.tecochat.chat.domain.ChatQueryRepository;
import chat.teco.tecochat.chat.domain.ChatQueryRepository.ChatSearchCond;
import chat.teco.tecochat.chat.domain.ChatRepository;
import chat.teco.tecochat.chat.domain.Question;
import chat.teco.tecochat.chat.domain.QuestionAndAnswer;
import chat.teco.tecochat.chat.domain.keyword.KeywordRepository;
import chat.teco.tecochat.chat.dto.ChatQueryDto;
import chat.teco.tecochat.chat.dto.ChatQueryDto.MessageQueryDto;
import chat.teco.tecochat.chat.dto.ChatSearchQueryDto;
import chat.teco.tecochat.chat.fixture.ChatFixture;
import chat.teco.tecochat.like.chatlike.domain.ChatLikeRepository;
import chat.teco.tecochat.member.domain.Course;
import chat.teco.tecochat.member.domain.Member;
import chat.teco.tecochat.member.domain.MemberRepository;
import java.util.List;
import java.util.Optional;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.DisplayNameGenerator.ReplaceUnderscores;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.support.PageableExecutionUtils;

@SuppressWarnings("NonAsciiCharacters")
@DisplayNameGeneration(ReplaceUnderscores.class)
@DisplayName("ChatQueryService 은(는)")
class ChatQueryServiceTest {

    private final ChatRepository chatRepository = mock(ChatRepository.class);
    private final MemberRepository memberRepository = mock(MemberRepository.class);
    private final ChatQueryRepository chatQueryRepository = mock(ChatQueryRepository.class);
    private final ChatLikeRepository chatLikeRepository = mock(ChatLikeRepository.class);
    private final KeywordRepository keywordRepository = mock(KeywordRepository.class);

    @InjectMocks
    private ChatQueryService chatQueryService = new ChatQueryService(
            memberRepository, chatRepository, chatQueryRepository, chatLikeRepository, keywordRepository);

    @Test
    void 단일_채팅_기록을_전부_조회한다() {
        // given
        final Chat chat = ChatFixture.chat(
                new QuestionAndAnswer(
                        question("안녕"),
                        answer("안녕하세요"),
                        2
                ),
                new QuestionAndAnswer(
                        Question.question("반가워"),
                        Answer.answer("저도 반갑습니다"),
                        2
                ));
        given(chatRepository.findById(1L))
                .willReturn(Optional.of(chat));
        given(memberRepository.findById(any()))
                .willReturn(Optional.of(new Member("말랑", Course.BACKEND)));

        // when
        final ChatQueryDto chatQueryDto = chatQueryService.findById(1L, 1L);

        // then
        assertAll(
                () -> assertThat(chatQueryDto.crewName()).isEqualTo("말랑"),
                () -> assertThat(chatQueryDto.title()).isEqualTo("안녕"),
                () -> assertThat(chatQueryDto.messages())
                        .extracting(MessageQueryDto::content)
                        .containsExactly("안녕", "안녕하세요", "반가워", "저도 반갑습니다")
        );
    }

    @Test
    void 검색할_수_있다() {
        // given
        final Chat chat1 = ChatFixture.chat(
                new QuestionAndAnswer(
                        question("안녕"),
                        answer("안녕하세요"),
                        2
                ));

        final Chat chat = ChatFixture.chat(
                new QuestionAndAnswer(
                        question("안녕2"),
                        answer("안녕하세요2"),
                        2
                ));
        final Page<Chat> page = PageableExecutionUtils.getPage(
                List.of(chat1, chat),
                PageRequest.of(0, 10),
                () -> 0);
        given(chatQueryRepository.search(any(), any()))
                .willReturn(page);
        given(memberRepository.findById(any()))
                .willReturn(Optional.of(new Member("말랑", Course.BACKEND)));

        // when
        final Page<ChatSearchQueryDto> search = chatQueryService.search(
                new ChatSearchCond(null, null, null),
                PageRequest.of(1, 1)
        );

        // then
        final ChatSearchQueryDto chatSearchQueryDto = search.getContent().get(0);
        assertAll(
                () -> assertThat(chatSearchQueryDto.crewName()).isEqualTo("말랑"),
                () -> assertThat(chatSearchQueryDto.course()).isEqualTo(Course.BACKEND),
                () -> assertThat(chatSearchQueryDto.title()).isEqualTo("안녕")
        );
    }
}

File: src/test/kotlin/chat/teco/tecochat/application/ChatQueryServiceTest.kt
package chat.teco.tecochat.application

import chat.teco.tecochat.createChat
import chat.teco.tecochat.createChatLike
import chat.teco.tecochat.createMember
import chat.teco.tecochat.createQuestionAndAnswer
import chat.teco.tecochat.domain.chat.ChatRepository
import chat.teco.tecochat.domain.chat.getByIdOrThrow
import chat.teco.tecochat.domain.chatlike.ChatLikeRepository
import chat.teco.tecochat.domain.keyword.KeywordRepository
import chat.teco.tecochat.domain.member.Course
import chat.teco.tecochat.domain.member.MemberRepository
import chat.teco.tecochat.query.ChatQueryRepository
import chat.teco.tecochat.query.ChatSearchCond
import chat.teco.tecochat.query.SearchChatResponse
import io.kotest.assertions.assertSoftly
import io.kotest.assertions.extracting
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.support.PageableExecutionUtils

class ChatQueryServiceTest : StringSpec({
    val memberRepository = mockk<MemberRepository>()
    val chatRepository = mockk<ChatRepository>()
    val chatQueryRepository = mockk<ChatQueryRepository>()
    val chatLikeRepository = mockk<ChatLikeRepository>()
    val keywordRepository = mockk<KeywordRepository>()

    val chatQueryService = ChatQueryService(
        memberRepository,
        chatRepository,
        chatQueryRepository,
        chatLikeRepository,
        keywordRepository
    )

    "채팅을 단일 조회한다" {
        val member = createMember(name = "herb", id = 1L)
        val chat = createChat(
            memberId = member.id,
            questionAndAnswers = listOf(
                createQuestionAndAnswer("질문1", "답변1"),
                createQuestionAndAnswer("질문2", "답변2"),
                createQuestionAndAnswer("질문3", "답변3")
            )
        )
        every { chatRepository.getByIdOrThrow(any()) } returns chat
        every { memberRepository.getById(any()) } returns member
        every { chatLikeRepository.findByMemberIdAndChatId(any(), any()) } returns createChatLike()
        every { keywordRepository.findAllByChatId(any()) } returns emptyList()

        val result = chatQueryService.findById(1L, 1L)

        assertSoftly(result) {
            it.crewName shouldBe "herb"
            it.course shouldBe Course.BACKEND
            it.likeCount shouldBe 0
            it.keywords shouldBe emptyList()
            it.title shouldBe chat.title
            it.isAlreadyClickLike shouldBe true
            extracting(it.messages) { content }
                .shouldContainExactly(
                    "질문1", "답변1", "질문2", "답변2", "질문3", "답변3"
                )
        }
    }

    "채팅을 검색한다" {
        val mallang = createMember(name = "mallang", id = 1L)
        val herb = createMember(name = "herb", id = 2L)
        val mallangChat = createChat(
            memberId = mallang.id,
            questionAndAnswers = listOf(createQuestionAndAnswer(), createQuestionAndAnswer())
        )
        val herbChat = createChat(
            memberId = herb.id,
            questionAndAnswers = listOf(createQuestionAndAnswer())
        )
        every { chatQueryRepository.search(any(), any()) } returns PageableExecutionUtils.getPage(
            listOf(mallangChat, herbChat), PageRequest.of(0, 10)
        ) { 0 }
        every { keywordRepository.findAllInChatIds(any()) } returns emptyList()
        every { memberRepository.findAllById(any()) } returns listOf(mallang, herb)

        val result: Page<SearchChatResponse> = chatQueryService.search(
            ChatSearchCond(null, null, null, null),
            PageRequest.of(1, 1)
        )

        assertSoftly(result.content[0]) {
            it.crewName shouldBe "mallang"
            it.course shouldBe Course.BACKEND
            it.commentCount shouldBe 0
            it.likeCount shouldBe 0
            it.keywords shouldBe emptyList()
            it.title shouldBe mallangChat.title
            it.totalQnaCount shouldBe 2
        }
        assertSoftly(result.content[1]) {
            it.crewName shouldBe "herb"
            it.course shouldBe Course.BACKEND
            it.commentCount shouldBe 0
            it.likeCount shouldBe 0
            it.keywords shouldBe emptyList()
            it.title shouldBe herbChat.title
            it.totalQnaCount shouldBe 1
        }
    }
})
